### ISSUE TITLE
Durable Function Fan-Out not Scaling Out (with Activity Triggers)

### ISSUE THREAD

**Issue by scott-aptuity** [Azure/azure-functions-host#4913]
Hi,
We are prototyping the standard Durable function Fan-Out scenario.
We have a loop that calls around 700 Activity Functions and then waits for them to complete.

My issue is that the activity functions don't appear to causing the functions to scale out - at least not any any reasonable amount of time.  

I see HTTP triggers will scale every 1 second, but other are 30 seconds?  
My Durable function is HTTP triggered and a I want to do a lot of identical work in parallel.  If the Fan-Out situation doesn't cause the function to scale, I'm not sure how useful it is.

Am I missing something or is Fan-Out not good at scaling out?

Thanks,
Scott

**Comment by brettsam** [comment]
Do you see zero scaleout -- or slower scaleout than you expect? The scaling logic will add instances when it sees that the backlog of messages is growing faster that it's able to keep up. Http scales much faster than other triggers, so that may be what you're experiencing. 

+@ConnorMcMahon / @cgillum who may have more to say. If you want to share your app name (either explicitly or [privately](https://github.com/Azure/azure-functions-host/wiki/Sharing-Your-Function-App-name-privately)), we may be able to see if there was any specific issue in your case.

**Comment by scott-aptuity** [comment]
I think its scaling out slower (30 second poll) than I would expect.  I would like to have the Activity Triggers scale out much quicker so the Fan Out can be done in parallel.  With the 30 second poll, the activity is basically being done serially since its not Scaling Out.  I was hoping it would Scale Out quickly and all the activities could be done in parallel.  This seems like a fairly typical Fan-Out scenario.

Each Fan Out Activity downloads a json file from blob and does some processing on it.  Nothing massive, maybe about 250ms total work time.    If we are doing 1000, I was hoping to speed it up with on demand resources.

**Comment by ChuckEdx** [comment]
I second this.

I'm also testing a durable function fan-out scenario. The orchestrator creates a blast of about ~300 activity functions and waits from them to complete. Each activity function takes around 4 seconds. What I end up seeing is the fan-out taking ~1200 seconds because the scale controller never adds instances. What I was hoping to see was the scale controller adding a bunch of instances to process this data as fast as possible, but unfortunately it's just happy to plug away with a single instance.

Is this because the activity function calls aren't piling up (because it's just a single blast of 300 and then nothing)?

Is there a way to tell the scale controller to be super aggressive in adding instances? I was hoping to see it just process the activity functions with 100 instances and thus be 100 times faster (ignoring overhead and whatnot).

**Comment by cgillum** [comment]
@scott-aptuity Yes, the scale controller currently scales out every 30 seconds in the Azure Functions Consumption plan. For Elastic Premium plans, it's once instance per 15 seconds. The reasons we originally had that disparity are no longer valid so we're looking into fixing soon so that non-HTTP workloads can scale as responsively as the HTTP workloads.

@ChuckEdx for your case it sounds like you're just not scaling at all, which isn't expected given the number of messages and the total time to process them all. The algorithm we use to scale Durable Function apps is pretty simple: has a message been sitting in the work-item queue waiting for more than 1 second? If so, scale out. That's obviously not happening for you so I think it would be good for us to investigate. If you have an Azure support plan, I suggest opening a case. Otherwise if you can share some information about your function app (the Azure region and the app name or some other unique identifier like an orchestration instance ID) and the timeframe, then we can take a quick look to try and figure it out.

(Also, I'm transferring this to the Durable Functions GitHub repo).
